File: malicious_client/src/NetworkMonitor.ts
import { EventEmitter } from 'events';
import * as os from 'os';
import * as fs from 'fs';
import { exec } from 'child_process';
import { promisify } from 'util';

const execAsync = promisify(exec);

export interface NetworkMetrics {
  downloadSpeed: number;
  uploadSpeed: number;
  latency: number;
  bandwidth: number;
  packetsPerSecond: number;
  networkInterface: string;
  connectionCount: number;
  packetLoss: number;
}

export class NetworkMonitor extends EventEmitter {
  private metrics: NetworkMetrics = {
    downloadSpeed: 0,
    uploadSpeed: 0,
    latency: 0,
    bandwidth: 0,
    packetsPerSecond: 0,
    networkInterface: '',
    connectionCount: 0,
    packetLoss: 0
  };

  private intervalId: NodeJS.Timeout | null = null;
  private lastCheck = Date.now();
  private lastNetworkStats: any = null;
  private primaryInterface: string = '';

  constructor() {
    super();
    this.detectPrimaryInterface();
    this.startMonitoring();
  }

  private async detectPrimaryInterface(): Promise<void> {
    try {
      const interfaces = os.networkInterfaces();
      
      // Find the primary network interface (first non-loopback, non-internal)
      for (const [name, addrs] of Object.entries(interfaces)) {
        if (addrs) {
          for (const addr of addrs) {
            if (!addr.internal && addr.family === 'IPv4') {
              this.primaryInterface = name;
              this.metrics.networkInterface = name;
              console.log(`Monitoring network interface: ${name} (${addr.address})`);
              return;
            }
          }
        }
      }
      
      // Fallback to eth0 or first available
      this.primaryInterface = Object.keys(interfaces)[0] || 'eth0';
      this.metrics.networkInterface = this.primaryInterface;
    } catch (error) {
      console.warn('Could not detect network interface, using eth0:', error);
      this.primaryInterface = 'eth0';
      this.metrics.networkInterface = 'eth0';
    }
  }

  private startMonitoring(): void {
    this.intervalId = setInterval(async () => {
      await this.updateRealMetrics();
      this.emit('metrics', this.metrics);
    }, 1000);
  }

  private async updateRealMetrics(): Promise<void> {
    try {
      await Promise.all([
        this.updateNetworkSpeed(),
        this.updateLatency(),
        this.updateConnectionCount(),
        this.updatePacketStats()
      ]);
    } catch (error) {
      console.warn('Error updating network metrics:', error);
      // Fall back to simulated data if real monitoring fails
      this.updateFallbackMetrics();
    }
  }

  private async updateNetworkSpeed(): Promise<void> {
    try {
      // Read network statistics from /proc/net/dev (Linux)
      if (process.platform === 'linux') {
        const data = await fs.promises.readFile('/proc/net/dev', 'utf8');
        const lines = data.split('\n');
        
        for (const line of lines) {
          if (line.includes(this.primaryInterface)) {
            const stats = line.trim().split(/\s+/);
            const rxBytes = parseInt(stats[1]) || 0;
            const txBytes = parseInt(stats[9]) || 0;
            const rxPackets = parseInt(stats[2]) || 0;
            const txPackets = parseInt(stats[10]) || 0;
            
            if (this.lastNetworkStats) {
              const now = Date.now();
              const elapsed = (now - this.lastCheck) / 1000;
              
              const rxDiff = rxBytes - this.lastNetworkStats.rxBytes;
              const txDiff = txBytes - this.lastNetworkStats.txBytes;
              const rxPacketDiff = rxPackets - this.lastNetworkStats.rxPackets;
              const txPacketDiff = txPackets - this.lastNetworkStats.txPackets;
              
              this.metrics.downloadSpeed = Math.max(0, rxDiff / elapsed);
              this.metrics.uploadSpeed = Math.max(0, txDiff / elapsed);
              this.metrics.packetsPerSecond = Math.max(0, (rxPacketDiff + txPacketDiff) / elapsed);
              this.metrics.bandwidth = this.metrics.downloadSpeed + this.metrics.uploadSpeed;
            }
            
            this.lastNetworkStats = { rxBytes, txBytes, rxPackets, txPackets };
            break;
          }
        }
      } else {
        // macOS/Windows - use netstat or similar
        await this.updateNetworkSpeedCrossPlatform();
      }
    } catch (error) {
      console.warn('Error reading network speed:', error);
    }
  }

  private async updateNetworkSpeedCrossPlatform(): Promise<void> {
    try {
      let command = '';
      
      if (process.platform === 'darwin') {
        // macOS
        command = `netstat -ibn | grep -E "^${this.primaryInterface}" | head -1`;
      } else if (process.platform === 'win32') {
        // Windows
        command = `wmic path Win32_NetworkAdapter where "Name like '%${this.primaryInterface}%'" get BytesReceivedPerSec,BytesSentPerSec /value`;
      }
      
      if (command) {
        const { stdout } = await execAsync(command);
        // Parse the output based on platform
        this.parseNetworkOutput(stdout);
      }
    } catch (error) {
      console.warn('Cross-platform network monitoring failed:', error);
    }
  }

  private parseNetworkOutput(output: string): void {
    // Implementation varies by platform
    // For now, use simulated data as fallback
    this.updateFallbackMetrics();
  }

  private async updateLatency(): Promise<void> {
    try {
      // Ping a reliable host to measure latency
      const command = process.platform === 'win32' 
        ? 'ping -n 1 8.8.8.8' 
        : 'ping -c 1 8.8.8.8';
        
      const { stdout } = await execAsync(command);
      
      // Extract latency from ping output
      const match = stdout.match(/time[<=](\d+(?:\.\d+)?)/i);
      if (match) {
        this.metrics.latency = parseFloat(match[1]);
      }
    } catch (error) {
      // If ping fails, estimate latency
      this.metrics.latency = 20 + Math.random() * 30; // 20-50ms
    }
  }

  private async updateConnectionCount(): Promise<void> {
    try {
      const command = process.platform === 'win32'
        ? 'netstat -an | find "ESTABLISHED" /c'
        : 'netstat -an | grep ESTABLISHED | wc -l';
        
      const { stdout } = await execAsync(command);
      this.metrics.connectionCount = parseInt(stdout.trim()) || 0;
    } catch (error) {
      this.metrics.connectionCount = Math.floor(Math.random() * 50 + 10); // 10-60 connections
    }
  }

  private async updatePacketStats(): Promise<void> {
    try {
      if (process.platform === 'linux') {
        // Read packet loss from /proc/net/snmp
        const data = await fs.promises.readFile('/proc/net/snmp', 'utf8');
        const lines = data.split('\n');
        
        for (const line of lines) {
          if (line.startsWith('Tcp:') && !line.includes('RtoAlgorithm')) {
            const stats = line.split(/\s+/);
            // Calculate packet loss percentage based on retransmissions
            const retransSegs = parseInt(stats[12]) || 0;
            const outSegs = parseInt(stats[11]) || 1;
            this.metrics.packetLoss = (retransSegs / outSegs) * 100;
            break;
          }
        }
      }
    } catch (error) {
      this.metrics.packetLoss = Math.random() * 2; // 0-2% packet loss
    }
  }

  private updateFallbackMetrics(): void {
    // Realistic fallback metrics when real monitoring isn't available
    const baseSpeed = 1000000; // 1MB/s base
    const variation = Math.random() * 0.5 + 0.75; // 75-125% variation
    
    this.metrics.downloadSpeed = baseSpeed * variation;
    this.metrics.uploadSpeed = this.metrics.downloadSpeed * (0.1 + Math.random() * 0.2);
    this.metrics.latency = 15 + Math.random() * 25; // 15-40ms
    this.metrics.bandwidth = this.metrics.downloadSpeed + this.metrics.uploadSpeed;
    this.metrics.packetsPerSecond = 200 + Math.random() * 300;
    this.metrics.connectionCount = Math.floor(Math.random() * 30 + 10);
    this.metrics.packetLoss = Math.random() * 1.5;
  }

  public recordTransfer(bytes: number, packets: number = 1): void {
    // This method can be used by the attack tool to record additional traffic
    this.metrics.downloadSpeed += bytes;
    this.metrics.packetsPerSecond += packets;
  }

  public getMetrics(): NetworkMetrics {
    return { ...this.metrics };
  }

  public stop(): void {
    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = null;
    }
  }
}

File: malicious_client/src/network_monitor.py
import os
import sys
import time
import threading
import psutil
import subprocess

class NetworkMetrics:
    def __init__(self):
        self.download_speed = 0.0
        self.upload_speed = 0.0
        self.latency = 0.0
        self.bandwidth = 0.0
        self.packets_per_second = 0.0
        self.network_interface = ""
        self.connection_count = 0
        self.packet_loss = 0.0

class NetworkMonitor:
    def __init__(self):
        self.metrics = NetworkMetrics()
        self.interval_id = None
        self.last_check = time.time()
        self.last_network_stats = None
        self.primary_interface = ""
        self._running = False
        self.detect_primary_interface()
        self.start_monitoring()

    def detect_primary_interface(self):
        # Pick first non-loopback, non-internal
        for iface, addrs in psutil.net_if_addrs().items():
            for addr in addrs:
                if addr.family == psutil.AF_LINK:
                    continue
                if getattr(addr, "address", "").startswith("127."):
                    continue
                self.primary_interface = iface
                self.metrics.network_interface = iface
                print(f"Monitoring network interface: {iface} ({addr.address})")
                return
        # Fallback to first available
        ifaces = list(psutil.net_if_addrs().keys())
        self.primary_interface = ifaces[0] if ifaces else 'eth0'
        self.metrics.network_interface = self.primary_interface

    def start_monitoring(self):
        self._running = True
        def loop():
            while self._running:
                self.update_real_metrics()
                time.sleep(1)
        self.interval_id = threading.Thread(target=loop)
        self.interval_id.daemon = True
        self.interval_id.start()

    def update_real_metrics(self):
        try:
            self.update_network_speed()
            self.update_latency()
            self.update_connection_count()
            self.update_packet_stats()
        except Exception as e:
            print("Error updating network metrics:", e)
            self.update_fallback_metrics()

    def update_network_speed(self):
        try:
            net_io = psutil.net_io_counters(pernic=True)
            iface = self.primary_interface
            now = time.time()
            if iface in net_io:
                stats = net_io[iface]
                rx_bytes = stats.bytes_recv
                tx_bytes = stats.bytes_sent
                rx_packets = stats.packets_recv
                tx_packets = stats.packets_sent
                if self.last_network_stats:
                    elapsed = now - self.last_check
                    rx_diff = rx_bytes - self.last_network_stats['rx_bytes']
                    tx_diff = tx_bytes - self.last_network_stats['tx_bytes']
                    rx_packet_diff = rx_packets - self.last_network_stats['rx_packets']
                    tx_packet_diff = tx_packets - self.last_network_stats['tx_packets']
                    self.metrics.download_speed = max(0, rx_diff / elapsed)
                    self.metrics.upload_speed = max(0, tx_diff / elapsed)
                    self.metrics.packets_per_second = max(0, (rx_packet_diff + tx_packet_diff) / elapsed)
                    self.metrics.bandwidth = self.metrics.download_speed + self.metrics.upload_speed
                self.last_network_stats = {
                    'rx_bytes': rx_bytes,
                    'tx_bytes': tx_bytes,
                    'rx_packets': rx_packets,
                    'tx_packets': tx_packets
                }
                self.last_check = now
        except Exception as e:
            print("Error reading network speed:", e)

    def update_latency(self):
        try:
            host = "8.8.8.8"
            count = 1
            if sys.platform == 'win32':
                command = ['ping', '-n', str(count), host]
            else:
                command = ['ping', '-c', str(count), host]
            result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
            output = result.stdout
            # Extract ms latency
            import re
            match = re.search(r"time[=<]([0-9.]+)", output)
            if match:
                self.metrics.latency = float(match.group(1))
        except Exception:
            self.metrics.latency = 20 + (30 * os.urandom(1)[0] / 255.0)

    def update_connection_count(self):
        try:
            # Count established connections
            conns = psutil.net_connections(kind="inet")
            self.metrics.connection_count = sum(1 for c in conns if c.status == "ESTABLISHED")
        except Exception:
            self.metrics.connection_count = int(10 + 50 * os.urandom(1)[0] / 255.0)

    def update_packet_stats(self):
        try:
            # Packet loss estimation: not directly available, simulate
            self.metrics.packet_loss = float("{:.2f}".format(os.urandom(1)[0] / 255.0 * 2))
        except Exception:
            self.metrics.packet_loss = float("{:.2f}".format(os.urandom(1)[0] / 255.0 * 2))

    def update_fallback_metrics(self):
        base_speed = 1000000  # 1MB/s
        import random
        variation = random.uniform(0.75, 1.25)
        self.metrics.download_speed = base_speed * variation
        self.metrics.upload_speed = self.metrics.download_speed * random.uniform(0.1, 0.3)
        self.metrics.latency = random.uniform(15, 40)
        self.metrics.bandwidth = self.metrics.download_speed + self.metrics.upload_speed
        self.metrics.packets_per_second = random.uniform(200, 500)
        self.metrics.connection_count = random.randint(10, 40)
        self.metrics.packet_loss = random.uniform(0, 1.5)

    def record_transfer(self, bytes_amount, packets=1):
        self.metrics.download_speed += bytes_amount
        self.metrics.packets_per_second += packets

    def get_metrics(self):
        return vars(self.metrics).copy()

    def stop(self):
        self._running = False

# Usage example:
# nm = NetworkMonitor()
# time.sleep(2)
# print(nm.get_metrics())
# nm.stop()
